### PR TITLE
Scoreboard

### PR DESCRIPTION
**Pull request by nwinter**
screenshot 2018-01-18 11 37 29

Can only test by setting a level assessment type to `open-ended` and adding a `scoreType` (will use the first one in the list). Here's a test level configured that way: https://staging.codecombat.com/play/level/desert-maze-open-scoreboard-test

### CODE REVIEW

File: app/models/LevelSession.coffee [hunk 1]
```diff
@@ -86,13 +86,19 @@ module.exports = class LevelSession extends CocoModel
...
         newTopScores.push type: scoreType, date: now, score: newScore
       else
         newTopScores.push oldTopScore
     state.topScores = newTopScores
     @set 'state', state
+    Backbone.Mediator.publish 'level:top-scores-updated', scores: @getTopScores()
+
+  getTopScores: ->
+    scores = (_.clone(score) for score in @get('state')?.topScores ? [])
+    score.score *= -1 for score in scores when score.type in ['time', 'damage-taken', 'code-length']  # Undo negative storage for display
```
Comment (sderickson) [inline app/models/LevelSession.coffee]: Should move this list out to a common variable in this file.

File: app/schemas/models/level_session.coffee [hunk 2]
```diff
@@ -136,7 +136,7 @@ _.extend LevelSessionSchema.properties,
         source: {type: 'string', enum: ['click']}  # Do not store 'code' flag events in the session.
     topScores: c.array {},
       c.object {},
-        type: c.shortString('enum': ['time', 'damage-taken', 'damage-dealt', 'gold-collected', 'difficulty'])
+        type: c.shortString('enum': ['time', 'damage-taken', 'damage-dealt', 'gold-collected', 'difficulty', 'code-length', 'survival-time', 'defeated'])
```
Comment (sderickson) [inline app/schemas/models/level_session.coffee]: Should share the score schema between here and level.

File: app/views/play/level/ControlBarView.coffee [hunk 3]
```diff
@@ -32,7 +38,7 @@ module.exports = class ControlBarView extends CocoView
     'click #control-bar-sign-up-button': 'onClickSignupButton'
     'click #version-switch-button': 'onClickVersionSwitchButton'
     'click #version-switch-button .code-language-selector': 'onClickVersionSwitchButton'
-    'click [data-toggle="coco-modal"][data-target="core/CreateAccountModal"]': 'openCreateAccountModal'
+    'click [data-toggle="coco-modal"][data-tarnnnnget="core/CreateAccountModal"]': 'openCreateAccountModal'
```
Comment (sderickson) [inline app/views/play/level/ControlBarView.coffee]: Keyboard fell asleep here? 😄

File: app/views/play/level/ControlBarView.coffee [hunk 4]
```diff
@@ -182,6 +189,51 @@ module.exports = class ControlBarView extends CocoView
     return if @session.get('state')?.difficulty is @lastDifficulty
     @render()
 
+  onScoresUpdated: (e) ->
+    #return unless @level.get('assessment') in ['open-ended']
+    mainScore = e.scores?[0]
```
Comment (sderickson) [inline app/views/play/level/ControlBarView.coffee]: If this is an expectation, this should be mentioned in the schema, and/or put in a separate property in the event for clarity.

File: app/views/play/level/ControlBarView.coffee [hunk 5]
```diff
@@ -182,6 +189,51 @@ module.exports = class ControlBarView extends CocoView
...
+      utils.replaceText $scoreboard.find('.best-score-value'), bestScoreText
+
+  formatScore: (type, score, isBest) ->
+    switch type
+      when 'time', 'survival-time'
+        if isBest
+          score.toFixed(1)
+        else
+          "#{score.toFixed(1)} #{$.i18n.t 'units.seconds'}"
+      else Math.round score
```
Comment (sderickson) [inline app/views/play/level/ControlBarView.coffee]: This logic in general could be put into the vuex store, and this piece of the front end could be replaced with a vue component. If you're feeling like digging in there!

Comment (nwinter) [reply app/views/play/level/ControlBarView.coffee]: You'll have to train me in the ways of the Vue sometime, maybe the next time I have a nice little new component to make.

**Pull request merged**